clients/src/main/java/org/apache/kafka/common/utils/Utils.java

```diff
@@ -784,4 +784,29 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
         return crc.getValue();
     }
 
+    /**
+     * Read data from the channel to the given byte buffer
+     * @param channel File channel containing the data to read from
+     * @param destinationBuffer The buffer into which bytes are to be transferred
+     * @param position The file position at which the transfer is to begin; must be non-negative
+     *
+     * @throws IllegalArgumentException If invalid arguments were passed
+     * @throws IOException If some other I/O error occurs
+     */
+    public static void readFully(FileChannel channel, ByteBuffer destinationBuffer, long position) throws IOException {
+        if (channel == null) {
+            throw new IllegalArgumentException("Cannot read an empty file channel.");
+        }
+        if (position < 0) {
+            throw new IllegalArgumentException("The file position cannot be negative, but we got " + position);
+        }
+        long currentPosition = position;
+        while (destinationBuffer.hasRemaining()) {
```

Comment:
I think we probably want a `do/while` loop here. There should be no difference in behaviour, but it seems to model the problem better (i.e. we first do a read and then we check if there is still space remaining in the buffer. Maybe:

```java
        long currentPosition = position;
        int bytesRead;
        do {
            bytesRead = channel.read(destinationBuffer, currentPosition);
            currentPosition += bytesRead;
        } while (bytesRead != -1 && destinationBuffer.hasRemaining());
```